admin_apps/journeys/iteration.py

```diff
@@ -447,67 +449,75 @@ def yaml_editor(yaml_str: str) -> None:
         language="yaml",
     )
 
-    button_container = st.container()
+    button_container = row(5, vertical_align="center")
     status_container_title = "**Edit**"
     status_container = st.empty()
 
-    with button_container:
-        (one, two, three, four) = st.columns(4)
-        if one.button("Validate", use_container_width=True, help=VALIDATE_HELP):
-            # Validate new content
-            try:
-                validate(
-                    content,
-                    snowflake_account=st.session_state.account_name,
-                    conn=get_snowflake_connection(),
-                )
-                st.session_state["validated"] = True
-                update_container(
-                    status_container, "success", prefix=status_container_title
-                )
-                st.session_state.semantic_model = yaml_to_semantic_model(content)
-                st.session_state.last_saved_yaml = content
-            except Exception as e:
-                st.session_state["validated"] = False
-                update_container(
-                    status_container, "failed", prefix=status_container_title
-                )
-                exception_as_dialog(e)
-
-            # Rerun the app if validation was successful.
-            # We shouldn't rerun if validation failed as the error popup would immediately dismiss.
-            # This must be done outside of the try/except because the generic Exception handling is catching the
-            # exception that st.rerun() properly raises to halt execution.
-            # This is fixed in later versions of Streamlit, but other refactors to the code are required to upgrade.
-            if st.session_state["validated"]:
-                st.rerun()
-
-        if content:
-            two.download_button(
-                label="Download",
-                data=content,
-                file_name="semantic_model.yaml",
-                mime="text/yaml",
-                use_container_width=True,
-                help=DOWNLOAD_HELP,
+    def validate_and_update_session_state() -> None:
```

Comment:
Diff in this file is a bit complicated to read, but I'm essentially just:

1. Moving the validation logic to a common function so that I can run it before the user clicks "Add joins", saving them a button click for validating
2. Adding a new button for "Add joins"